### PR TITLE
RUMM-2207 Delegate building HTTP request to features

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Ask Features for `URLRequest` at batch upload. At configuration, a Feature will provide an implementation of a `FeatureRequestBuilder` to build HTTP request for given ~payloads~ events and current core context.

### How?

All features now provide an implementation of:
```swift 
/// The `FeatureRequestBuilder` defines an interface for building a single `URLRequest`
/// for a list of data payloads and the current core context.
///
/// The core implementation can use the interface for creating requests targetting datadog intake
/// of a given feature.
internal protocol FeatureRequestBuilder {

    /// Builds a `URLRequest` for a list of payloads and the current core context to be uploaded
    /// to the feature intake.
    ///
    /// The returned request must include all necessary information, including HTTP headers and
    /// URL queries for the intake to ingest the payload. The request will be sent as is by the core
    /// uploader.
    ///
    /// - Parameters:
    ///   - context: The current core context.
    ///   - payloads: The payloads to upload.
    /// - Returns: The URL request.
    func request(for payloads: [Data], with context: /* DatadogContext */ DatadogV1Context) -> URLRequest
}
```
The feature's implementation of `FeatureRequestBuilder` will know how to format the data, decorate the HTTP Headers and query parameters, and provide the intake URL for the upload.

The `FileReader` not long needs to know the `DataFormat` and can directly return the list of event payloads.

#### Move files to `DatadogCore` and `DatadogInternal`
As part of this PR, I've started to move implementation to `DatadogCore` and `DatadogInternal`. Some object in the internal module have been marked with `/* public */` to show their futur access level.

#### Still V1 Context
This change still uses the `DatadogV1Context` to keep a low impact, but those references will soon be replaced with `DatadogContextProvider`.

#### Battery Status
In this PR, `BatteryStatus` is still private to the `DatadogCore` but I'm considering adding it the context.

#### `DatadogV1Context` in tests

`DatadogV1Context` keeps a reference to `CoreDependencies` which [keep a HTTP client of the `ServerMock` in tests](https://github.com/DataDog/dd-sdk-ios/blob/develop/Tests/DatadogTests/Datadog/Mocks/CoreMocks.swift#L503). This can easily lead to multiple interception and fail at [this precondition](https://github.com/DataDog/dd-sdk-ios/blob/develop/Tests/DatadogTests/Datadog/Mocks/ServerMock.swift#L161). I've added injection of the HTTPClient in mock, but I think will should soon move to the new context so we don't keep this tech debt too long.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests
- [x] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: Sources/Datadog/Core/Persistence/Reading/Reader.swift [hunk 1]
```diff
@@ -7,7 +7,7 @@
 import Foundation
 
 internal struct Batch {
-    let data: Data
+    let payloads: [Data]
```
Comment (ncreated) [inline Sources/Datadog/Core/Persistence/Reading/Reader.swift:10]: Similar comment on using "payloads" name. Here the problem is even more visible, as we're mixing domains - `Batch` doesn't need to know that it participates data uploads, so calling it `events: [Data]` could be even more desired here. I also wish we have a doc comment on this property as it's pretty crucial. WDYT?

**Review by ncreated — COMMENTED** [review]
Very nice work! 👍

Everything looks good to me, except the `payloads` name that I left comments on. I know it's "naming thing", but it is quite crucial element of the entire flow and IMO the way we use it leaves too much ambiguity (I had troubles to figure it out).

> Move files to DatadogCore and DatadogInternal

Sounds good 👍. I'm okay with using `/* public */` pattern to annotate "soon public" types. It's easier to find and remove in V2, than inline notes we also leave sometimes around.

> Still V1 Context

Totally OK 👍.

> Battery Status

Sounds good to include it in the context 👍. That would be almost free, and we were preliminary discussing some use of it in SR.

> DatadogV1Context in tests

The change on `HTTPClient` also looks fine 👍. Agree, it will be soon more limited and eventually gone. However, I still believe we'll need some tests in Core which use `ServerMock` (so mocked `HTTPProtocol`) - to assert data uploader.

File: Sources/Datadog/DatadogInternal/Upload/FeatureRequestBuilder.swift [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    ///
+    /// The returned request must include all necessary information, including HTTP headers and
+    /// URL queries for the intake to ingest the payload. The request will be sent as is by the core
+    /// uploader.
+    ///
+    /// - Parameters:
+    ///   - context: The current core context.
+    ///   - payloads: The payloads to upload.
+    /// - Returns: The URL request.
+    func request(for payloads: [Data], with context: /* DatadogContext */ DatadogV1Context) -> URLRequest
```
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/DatadogInternal/Upload/FeatureRequestBuilder.swift:26]: Yes, changing it to `events` sounds good to me 👍 I wasn't very satisfied with this name either, I just didn't want to keep `data`. And based on the IETF, `payload` doesn't match here!

File: Sources/Datadog/DatadogInternal/Upload/URLRequestBuilder.swift [hunk 3]
```diff
@@ -38,68 +24,64 @@ internal struct RequestBuilder {
         static let ddEVPOriginVersionHeaderField = "DD-EVP-ORIGIN-VERSION"
         static let ddRequestIDHeaderField = "DD-REQUEST-ID"
 
-        enum Value {
-            /// If the header's value is constant.
-            case constant(_ value: String)
-            /// If the header's value is different each time.
-            case dynamic(_ value: () -> String)
```
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/DatadogInternal/Upload/URLRequestBuilder.swift:45]: I was aiming for simplicity in the way we process the headers, and to allocate a single header dictionary. But the behaviour is the same, with contant vs. computed in closure 👍

**Comment by maxep** [comment]
It's green on [CI](https://app.bitrise.io/build/ed5909e8-0731-4a87-92a6-7ffa9a746965)..
Rebuilding to get the right status!

**Comment by maxep** [comment]
@ncreated 
> However, I still believe we'll need some tests in Core which use ServerMock (so mocked HTTPProtocol) - to assert data uploader.

Totally, I meant removing the `HTTPClient` dependency in context!

**Review by ncreated — APPROVED** [review]
Perfect 🚀

**Pull request merged**